src/backend/distributed/transaction/distributed_deadlock_detection.c

```diff
@@ -25,11 +25,247 @@
+		bool deadlockFound = false;
+		TransactionNode *transactionNode =
+			(TransactionNode *) hash_search(adjacencyList, transactionId, HASH_FIND,
+											&transactionFound);
+
+		Assert(transactionFound);
+
+		ResetVisitedFields(adjacencyList);
+
+		deadlockFound = CheckDeadlockForDistributedTransaction(transactionNode);
+		if (deadlockFound)
+		{
+			AssicateDistributedTransactionWithBackendProc(transactionNode);
+
+			KillBackendDueToDeadlock(transactionNode->initiatorProc);
+
+			return true;
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * CheckDeadlockForDistributedTransaction gets a transaction node and returns
+ * true if the given transaction participates in a distributed transaction.
+ *
+ * In essence, the function does a DFS starting with the given transaction node
+ * and checks for a cycle. Finding a cycle indicates a distributed deadlock and
+ * the function returns true on that case.
+ */
+static bool
+CheckDeadlockForDistributedTransaction(TransactionNode *transactionNode)
+{
+	List *waitingTransactionNodes = transactionNode->waitsFor;
+
+	while (waitingTransactionNodes != NIL)
+	{
+		TransactionNode *waitingTransactionNode =
+			(TransactionNode *) linitial(waitingTransactionNodes);
+		ListCell *currentWaitForCell = NULL;
+
+		waitingTransactionNodes = list_delete_first(waitingTransactionNodes);
+
+		/* cycle found, let the caller know about the cycle */
+		if (waitingTransactionNode == transactionNode)
+		{
+			return true;
+		}
+
+		/* don't need to revisit the node again */
+		if (waitingTransactionNode->transactionVisited)
+		{
+			continue;
+		}
+
+		waitingTransactionNode->transactionVisited = true;
+
+		/* prepend to the list to continue depth-first search */
+		foreach(currentWaitForCell, waitingTransactionNode->waitsFor)
+		{
+			TransactionNode *waitForTransaction =
+				(TransactionNode *) lfirst(currentWaitForCell);
+
+			waitingTransactionNodes =
+				lcons(waitForTransaction, waitingTransactionNodes);
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * ResetVisitedFields goes over all the elements of the input adjacency list
+ * and sets transactionVisited to false.
+ */
+static void
+ResetVisitedFields(HTAB *adjacencyList)
+{
+	HASH_SEQ_STATUS status;
+	TransactionNode *resetNode = NULL;
+
+	/* reset all visited fields */
+	hash_seq_init(&status, adjacencyList);
+
+	while ((resetNode = (TransactionNode *) hash_seq_search(&status)) != 0)
+	{
+		resetNode->transactionVisited = false;
+	}
+}
+
+
+/*
+ * AssicateDistributedTransactionWithBackendProc gets a transaction node
+ * and searches the corresponding backend. Once found, transactionNodes'
+ * initiatorProc is set to it.
+ *
+ * The function goes over all the backends, checks for the backend with
+ * the same transaction number as the given transaction node.
+ */
+static void
+AssicateDistributedTransactionWithBackendProc(TransactionNode *transactionNode)
+{
+	int backendIndex = 0;
+
+	for (backendIndex = 0; backendIndex < MaxBackends; ++backendIndex)
+	{
```

Comment:
This seems like a really expensive approach. Why do we need this? In my prototype that certainly wasn't necessary?  We already have the PID in the earlier represenation, we should just avoid throwing it away...